trpcultivate_phenotypes/tests/src/Kernel/Validators/MetadataInputTest.php

```diff
@@ -0,0 +1,180 @@
+<?php
+
+/**
+ * @file
+ * Kernel tests for validator plugins specific to validating metadata/input to importer.
+ */
+
+namespace Drupal\Tests\trpcultivate_phenotypes\Kernel\Validators;
+
+use Drupal\Tests\tripal_chado\Kernel\ChadoTestKernelBase;
+use Drupal\Tests\trpcultivate_phenotypes\Traits\PhenotypeImporterTestTrait;
+use Drupal\Core\Form\FormState;
+
+ /**
+  * Tests Tripal Cultivate Phenotypes Metadata Validator Plugins.
+  *
+  * @group trpcultivate_phenotypes
+  * @group validators
+  */
+class MetadataInputTest extends ChadoTestKernelBase {
+  use PhenotypeImporterTestTrait;
+
+  /**
+   * A genus that exists and is configured. 
+   */
+  protected $test_genus;
+
+  /**
+   * Modules to enable.
+   */
+  protected static $modules = [
+    'user',
+    'tripal',
+    'tripal_chado',
+    'trpcultivate_phenotypes'
+  ];
+
+  /**
+   * {@inheritdoc}
+   */
+  protected function setUp(): void {
+    parent::setUp();
+
+    // Set test environment.
+    \Drupal::state()->set('is_a_test_environment', TRUE);
+
+    // Install module configuration.
+    $this->installConfig(['trpcultivate_phenotypes']);
+
+    // Test Chado database.
+    // Create a test chado instance and then set it in the container for use by our service.
+    $this->connection = $this->createTestSchema(ChadoTestKernelBase::PREPARE_TEST_CHADO);
+    $this->container->set('tripal_chado.database', $this->connection);
+
+    // Set plugin manager service.
+    $this->plugin_manager = \Drupal::service('plugin.manager.trpcultivate_validator');
+
+    $genus = 'Tripalus';
+    // Create our organism and configure it.
+    $organism_id = $this->connection->insert('1:organism')
+      ->fields([
+        'genus' => $genus,
+        'species' => 'databasica',
+      ])
+      ->execute();
+
+    $this->assertIsNumeric($organism_id, 'We were not able to create an organism for testing (configured).');
+    $this->test_genus['configured'] = $genus;
+    $this->setOntologyConfig($this->test_genus['configured']);
+
+    // Create another organism but create a configuration item
+    // where the cv for trait is set to 0 - not configured.
+    $genus = 'notconfiggenus';
+    $organism_id = $this->connection->insert('1:organism')
+      ->fields([
+        'genus' => $genus,
+        'species' => 'databasica',
+      ])
+    ->execute();
+
+    $this->assertIsNumeric($organism_id, 'We were not able to create an organism for testing (not configured).');
+    $this->test_genus['not-configured'] = $genus;
+    $config = \Drupal::configFactory()->getEditable('trpcultivate_phenotypes.settings')
+      ->set('trpcultivate.phenotypes.ontology.cvdbon.' . $this->test_genus['not-configured'], ['trait' => 0]);
```

Comment:
This should not be needed. The checking of whether a genus is configured or not should not make the assumption that this exists as there are edge cases when the genus is added later where it would not be.

Reply:
not quite. in the real world, upon install the module will create a config for all genus and set all cv's to 0 - the not configured genus. Other genus added later will not have this configuration setting until set in the config page. Since the install routine was skipped in this test, non configured genus does not have that config item (set to 0) and is always treated as genus does not exists. This will a true test for genus where it was not configured yet.

Reply:
Exactly, "Other genus added later will not have this configuration setting until set in the config page." So if a genus is added and then this validator is triggered without anyone going to a config page then the module will not have this key set. Again, I just really like tests to make as few assumptions as possible and this feels like an assumption to me 😬 Also, I think it's unnecessary as I believe your validator will catch in even without you setting this in the test. If you want to be really thorough then I suggest adding another case where there is a genus without this set that exists but is not configured. That way you check my case and yours and we will know if the validator works even if your assumption is not true in some weird edge case.